core/src/main/scala/akka/persistence/r2dbc/internal/codec/TimestampCodec.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2022 - 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.r2dbc.internal.codec
+
+import java.time.Instant
+import java.time.LocalDateTime
+import java.util.TimeZone
+
+import io.r2dbc.spi.Row
+import io.r2dbc.spi.Statement
+
+import akka.annotation.InternalApi
+import akka.persistence.r2dbc.internal.InstantFactory
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] sealed trait TimestampCodec {
+
+  def encode(timestamp: Instant): Any
+  def decode(row: Row, name: String): Instant
+  def decode(row: Row, index: Int): Instant
+
+  def instantNow(): Instant = InstantFactory.now()
+
+  def now[T](): T
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] object TimestampCodec {
+
+  class PostgresTimestampCodec extends TimestampCodec {
+    override def decode(row: Row, name: String): Instant = row.get(name, classOf[Instant])
+    override def decode(row: Row, index: Int): Instant = row.get(index, classOf[Instant])
+
+    override def encode(timestamp: Instant): Any = timestamp
+
+    override def now[T](): T = instantNow().asInstanceOf[T]
+  }
+  object PostgresTimestampCodec extends PostgresTimestampCodec
+
+  case object SqlServerTimestampCodec extends TimestampCodec {
+
+    // should this come from config?
+    private val zone = TimeZone.getTimeZone("UTC").toZoneId
+
+    private def toInstant(timestamp: LocalDateTime) =
+      timestamp.atZone(zone).toInstant
+
+    override def decode(row: Row, name: String): Instant = toInstant(row.get(name, classOf[LocalDateTime]))
+
+    override def encode(timestamp: Instant): LocalDateTime = LocalDateTime.ofInstant(timestamp, zone)
+
+    override def now[T](): T = LocalDateTime.ofInstant(instantNow(), zone).asInstanceOf[T]
```

Comment:
that cast to `T` looks dangerous, for example this usage:
```
override def currentDbTimestamp(): Future[Instant] = Future.successful(timestampCodec.now())
```

wouldn't that cast a LocalDateTime to an Instant?

Reply:
Yes, right. I am fixing this now. It seems I can get rid of `def now[T](): T` completely, and fix the places where I am not yet binding a timestamp with the codec specific `.bindTimestamp`.

The unsafe casts must not yet have been triggered I assume.